mathics/builtin/string/operations.py

```diff
@@ -45,6 +47,96 @@
 SymbolAll = Symbol("All")
 
 
+class _ZLibHash:  # make zlib hashes behave as if they were from hashlib
+    def __init__(self, fn):
+        self._bytes = b""
+        self._fn = fn
+
+    def update(self, bytes):
+        self._bytes += bytes
+
+    def hexdigest(self):
+        return format(self._fn(self._bytes), "x")
+
+
+class Hash(Builtin):
+    """
+    <dl>
+      <dt>'Hash[$expr$]'
+      <dd>returns an integer hash for the given $expr$.
+
+      <dt>'Hash[$expr$, $type$]'
+      <dd>returns an integer hash of the specified $type$ for the given $expr$.</dd>
+      <dd>The types supported are "MD5", "Adler32", "CRC32", "SHA", "SHA224", "SHA256", "SHA384", and "SHA512".</dd>
+
+      <dt>'Hash[$expr$, $type$, $format$]'
+      <dd>Returns the hash in the specified format.</dd>
+    </dl>
+
+    > Hash["The Adventures of Huckleberry Finn"]
+    = 213425047836523694663619736686226550816
+
+    > Hash["The Adventures of Huckleberry Finn", "SHA256"]
+    = 95092649594590384288057183408609254918934351811669818342876362244564858646638
+
+    > Hash[1/3]
+    = 56073172797010645108327809727054836008
+
+    > Hash[{a, b, {c, {d, e, f}}}]
+    = 135682164776235407777080772547528225284
+
+    > Hash[SomeHead[3.1415]]
+    = 58042316473471877315442015469706095084
+
+    >> Hash[{a, b, c}, "xyzstr"]
+     = Hash[{a, b, c}, xyzstr, Integer]
+    """
+
+    rules = {
+        "Hash[expr_]": 'Hash[expr, "MD5", "Integer"]',
+        "Hash[expr_, type_String]": 'Hash[expr, type, "Integer"]',
+    }
+
+    attributes = protected | read_protected
+
+    summary_text = "compute hash codes for a string"
+
+    # FIXME md2
+    _supported_hashes = {
+        "Adler32": lambda: _ZLibHash(zlib.adler32),
+        "CRC32": lambda: _ZLibHash(zlib.crc32),
+        "MD5": hashlib.md5,
+        "SHA": hashlib.sha1,
+        "SHA224": hashlib.sha224,
+        "SHA256": hashlib.sha256,
+        "SHA384": hashlib.sha384,
+        "SHA512": hashlib.sha512,
+    }
+
+    @staticmethod
+    def compute(user_hash, py_hashtype, py_format):
+        hash_func = Hash._supported_hashes.get(py_hashtype)
+        if hash_func is None:  # unknown hash function?
+            return  # in order to return original Expression
+        h = hash_func()
+        user_hash(h.update)
+        res = h.hexdigest()
+        if py_format in ("HexString", "HexStringLittleEndian"):
+            return String(res)
+        res = int(res, 16)
+        if py_format == "DecimalString":
+            return String(str(res))
+        elif py_format == "ByteArray":
+            return from_python(bytearray(res))
```

Comment:
from_python can be replaced by an Expression.

Reply:
@TiagoCavalcante Please just go ahead and make the changes you suggest. Thanks.